plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1530,6 +1542,8 @@ private void commitShared()
             // Clean up staging directories (that may recursively contain empty directories or stale files from failed attempts)
             committer.pruneAndDeleteStagingDirectories(declaredIntentionsToWrite);
         }
+
+        committer.executeDirectoryCallback();
```

Comment:
The callback won't be reached if any of the above fails, and we end up with cache desync'd from actual storage state. Did you consider calling invalidation after operations that change the state, instead of accumulating them first?

Reply:
This is a valid point.
I have added the logic related to dropping tables/partitions within the callable function performing the irreversible operation.


Regarding other changes which can happen due to DML operations/ ALTER TABLE operations, I think that the cache invalidation should be performed  also when performing abort operations.

> Did you consider calling invalidation after operations that change the state ?
Some operations are happening async. See `io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore.Committer#executeRenameTasksForAbort` 
I find that the maintenance of the performing the callback logic is easier when we record the resources being changed and then perform the callback while committing.


> The callback won't be reached if any of the above fails, and we end up with cache desync'd from actual storage state.
This is unfortunately true. Because the operations don't have in an atomic fashion there is the danger of performing operations that change the file system and fail in between leaving the cache not updated.
I added a routine to be called on doing rollbacks for tables/partitions, but yes, there is no 100% warranty that this routine will actually be reached by the code flow in case that a major problem occurs